### DISCUSSION TITLE
Using Adorners on the buttons

### DISCUSSION THREAD

**Discussion by Taldrit78** [fluentribbon/Fluent.Ribbon#1172] [Q&A] [answered]
We use Fluent.Ribbon now for some time and are happy with it, but now we got a requirement to show a notification element on the button (for now only a red dot, but it could become a counter or something the like).
For this we wanted to use the AdornerLayer and paint it over the existing one.
But we cannot get it to work.

First we create the buttons "by hand", like this:

    Button button = new Button()
    {
        KeyTip = buttonData.KeyTip
    };
    button.SetBinding(Button.HeaderProperty, headerBinding);
    button.SetBinding(Button.ToolTipProperty, tooltipBinding);
    button.SetBinding(Button.IconProperty, new Binding("Icon") { Source = buttonData });
    button.SetBinding(Button.LargeIconProperty, new Binding("LargeIcon") { Source = buttonData });
    button.SetBinding(Button.CommandProperty, new Binding("Command") { Source = buttonData });
    button.SetBinding(Button.StyleProperty, new Binding("Style") { Source = buttonData });
    button.SetBinding(Button.VisibilityProperty, new Binding("IsVisible") { Source = buttonData, Converter = new BooleanToVisibilityConverter() });
    buttonData.Control = button;
    groupBox.Items.Add(button);

Normally I would use something like this:

    AdornerLayer adornerLayer = AdornerLayer.GetAdornerLayer(button);
    adornerLayer.Add(new AdornerDot(button));
    adornerLayer.SetBinding(AdornerLayer.VisibilityProperty, buttonData.Highlight);

But as long as the button is not initialized inside the window, the VisualTree does not exist and the AdornerLayer does not exist, either.
The buttons are later not easily reachable by code-behind and because the ControlLayout of the buttons are hidden somewhere inside the Framework I cannot manipulate them.

Has anyone a good idea of how to put the adorner there without waiting for the window to have initialized first and then walking the VisualTree down till I find the correct button to put my Adorner on?

**Comment by batzen** [comment]
I also had to solve this kind of issue, getting adorners to show on any element.

You can either use an attached property that, when set to true, subscribes to the loaded and unloaded event of the control the property was used on and show/hide the adorner when the element gets loaded.

Or you could use a behavior that takes care of that.

The way with the most control is a combination of both.
A behavior that's responsible for showing/hiding the adorner and an attached property to attach the behavior.
Going only the behavior route prevents you from using it in styles, except you use something like `StylizedBehaviors` but that one also has it's disadvantages as the way you'd attach the behavior would differ between regular control usage and style definition.

**Comment by Taldrit78** [comment, marked as answer]
Ok, I found my solution...
If I create a AdornerDecorator and put the Button inside and THEN try to get the AdornerLayer I can manipulate it:

                                Button button = new Button()
                                {
                                    KeyTip = buttonData.KeyTip
                                };
                                AdornerDecorator adornerDecorator = new AdornerDecorator
                                {
                                    Child = button
                                };

                                if (buttonData.Highlight != null)
                                {
                                    AdornerLayer adornerLayer = AdornerLayer.GetAdornerLayer(button);
                                    adornerLayer.Add(new AdornerDot(button));
                                    adornerLayer.SetBinding(AdornerLayer.VisibilityProperty, buttonData.Highlight);
                                }
                                button.SetBinding(Button.HeaderProperty, headerBinding);
                                button.SetBinding(Button.ToolTipProperty, tooltipBinding);
                                button.SetBinding(Button.IconProperty, new Binding("Icon") { Source = buttonData });
                                button.SetBinding(Button.LargeIconProperty, new Binding("LargeIcon") { Source = buttonData });
                                button.SetBinding(Button.CommandProperty, new Binding("Command") { Source = buttonData });
                                button.SetBinding(Button.StyleProperty, new Binding("Style") { Source = buttonData });
                                button.SetBinding(Button.VisibilityProperty, new Binding("IsVisible") { Source = buttonData, Converter = new BooleanToVisibilityConverter() });

                                buttonData.Control = button;

                                groupBox.Items.Add(adornerDecorator);

Just have to add the AdornerDecorator to the group and the button as child of the decorator. This way it works fine.

**Reply by batzen** [reply]
Glad you found a solution that works for you.
